### ISSUE TITLE
Ability to link between refman and stdlib doc.

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#10142]
Sometimes, the reference manual refers to objects defined in the `.v` files. In this case, it would be good if we could link to their coqdoc-generated documentation. The idea of @cpitclaudel is that we would add a new inline role like ``` :stdlib:`Coq.Numbers.NatInt.NZAdd#NZAddProp.add_shuffle0` ``` which would link to `stdlib-doc-path/Coq.Numbers.NatInt.NZAdd.html#NZAddProp.add_shuffle0`.

Note that this would have to be a relative link, so that locally installed documentation doesn't link to the web, and so that documentation for `v8.X` doesn't link to documentation for `v8.Y`. This therefore requires to solve #8524 first. To be consistent with http://coq.github.io/doc/master/refman/ being the root for the HTML documentation of the manual and https://coq.github.io/doc/master/stdlib/ being the root for the HTML documentation of the stdlib (which is also consistent BTW with http://coq.inria.fr/refman/ and http://coq.inria.fr/stdlib/, https://coq.inria.fr/distrib/V8.8.2/refman/ and https://coq.inria.fr/distrib/V8.8.2/stdlib/, etc.), the natural solution to #8524 is that the HTML documentation be installed under `share/doc/coq/html/` in two folders: `refman` and `stdlib`.

Note also that it is probably going to be impossible to achieve the same thing in PDF versions of the documentation.

**Comment by cpitclaudel** [comment]
> Note also that it is probably going to be impossible to achieve the same thing in PDF versions of the documentation.

Why not use an absolute link to the correct online version of the docs? (even in HTML, the relative link is a convenience when the doc is installed locally, but otherwise shouldn't it be possible to link to a specific version of the manual?)

**Comment by Zimmi48** [comment]
> shouldn't it be possible to link to a specific version of the manual?

Yes, although you need to know which version of the manual to link to. Consider a CI build with some HTML artifact. It would be very convenient to be able to click a link and be taken to the documentation at the same version.

**Comment by cpitclaudel** [comment]
> Yes, although you need to know which version of the manual to link to. 

Good point.  Then maybe the right way to do this is to add a function to compute the current "link prefix".  The HTML one would return `""`, and the latex one would return the URL of the coq website, at the "right" version (which would only be right for releases, not for CI builds, but that would be better than nothing)?

**Comment by jfehrle** [comment]
I suggest that:
- We should ensure that you can easily build updated documentation with working links to stdlib.  This is to facilitate editing the doc and its links to stdlib--if it's cumbersome we'll never get them right.
- There should be a way to check at build time that links to stdlib are valid,  Not sure how you'd do that

**Comment by cpitclaudel** [comment]
> We should ensure that you can easily build updated documentation with working links to stdlib.

I have this working (mostly) in Alectryon.  I can port it to coqrst or we can wait to switch to Alectryon.

> There should be a way to check at build time that links to stdlib are valid, Not sure how you'd do that

You can either get coqdoc to print all the link targets it created somewhere, or you can open the HTML files and check directly.  The second one requires no coqdoc modifications, which is a plus.

**Comment by jfehrle** [comment]
> I have this working (mostly) in Alectryon. I can port it to coqrst or we can wait to switch to Alectryon.

Wouldn't it be possible (and simple?) to have a sphinx build flag that generates relative links to stdlib?
(I have no knowledge of or opinion on Alectryon.)

**Comment by cpitclaudel** [comment]
>  Wouldn't it be possible (and simple?) to have a sphinx build flag that generates relative links to stdlib?

Maybe! :) Can you explain what you mean in more detail?

**Comment by jfehrle** [comment]
If you look at https://coq.inria.fr/distrib/current/refman/language/extensions/evars.html, the link in the table of contents to "Sorts" under "Core Language" is `<a class="reference internal" href="../core/sorts.html">Sorts</a>` and it takes you to https://coq.inria.fr/distrib/current/refman/language/core/sorts.html.  (This is easy to see using the browser's HTML debugger; if you're not familiar with it I can elaborate for Chrome.)  The links are relative to the base directory https://coq.inria.fr/distrib/current/refman/.

Two ways to have links to stdlib work locally would be:
- adjust the makefile/Dune to generate the stdlib pages under https://coq.inria.fr/distrib/current/refman/.  Then get sphinx to output the right `href=` for stdlib links somehow TBD.
- leave the stdlib page output where it is but use an environment variable set by, say, `./configure -local` to point to the right directory.  Somehow TBD have Sphinx output the right `href=` (i.e, a relative path from the common superdirectory).

To go to a particular spot on a page, append the tag name after `#` as usual, e.g. `href="../../addendum/universe-polymorphism.html#grammar-token-universe-name"`

We ought to have a target for building the stdlib doc.  I don't know what it is.

Note the path names are indeed relative, so the links associated with the table of contents entries vary depending on what page you're on.

Kapish?  Questions?

**Comment by cpitclaudel** [comment]
> This is easy to see using the browser's HTML debugger; if you're not familiar with it I can elaborate for Chrome.

Very kind to offer :) I'm quite familiar with these tools ^^

> Then get sphinx to output the right href= for stdlib links somehow TBD.
>
> Somehow TBD have Sphinx output the right href= (i.e, a relative path from the common superdirectory).

Oh, that TBD is the only part I was discussing here.

**Comment by jfehrle** [comment]
Oh, I thought you know the Sphinx better than I do.

You could output a raw html node with the right info for a `:stdlib:` directive, bypassing the complications of `nodes.target`.  If we want index the `stdlib` references, we'd create both a raw node for the stdlib link and a regular target for use by the index.

pdf can't support external links, so no need to worry about that.

What do you think?

**Comment by cpitclaudel** [comment]
> Oh, I thought you know Sphinx better than I do.

:shrug: 

> What do you think?

I think we may be misunderstanding each other, and it's probably my fault; sorry about that :/

What you describe is mostly what I have in Alectryon, which is what I meant when I said "*I have this working (mostly) in Alectryon. I can port it to coqrst*".  I think I misunderstood the part you were referring to in your earlier message and that's what led to the confusion.  When you wrote "*Wouldn't it be possible (and simple?) to have a sphinx build flag that generates relative links to stdlib?*", I misunderstood which message you were referring to.  I think the answer should just have been "yes, it should be easy".

**Comment by SkySkimmer** [comment]
What's alectryon?

**Comment by Zimmi48** [comment]
Alectryon is a tool, inspired by coqrst but relying on SerAPI instead of calling `coqtop` directly, that @cpitclaudel has developed and that is used in the MIT blog announced in https://coq.discourse.group/t/plv-has-a-new-blog/925/3, but which is much more powerful than this example shows. It is currently under review as a conference paper and I think this is why Clément hasn't announced it publicly yet.

**Comment by Zimmi48** [comment]
Regarding the relative link discussion, my opinion on this is that we should always assume that the HTML documentation is installed in the same root folder, the refman in the `refman` subfolder and the stdlib doc in the `stdlib` subfolder. It used to be the case before the Sphinx migration but it is still a TODO to fix this in the Makefile. However, this convention is already respected when we deploy to the web.

**Comment by jfehrle** [comment]
@Zimmi48 That's good for when it's installed, but we ought to make it work with `./configure -local` too.
